### PR TITLE
build: Enables FreeBSD support

### PR DESCRIPTION
**Pull request by jasonwilliams200OK**
Changed hardcoded bash shebang to env to support multiple directory
structures (required to build on FreeBSD).

Related-PR: #1756

### CODE REVIEW

**Comment by josteink** [comment]
Not sure why CI fails. Looks much more complete and I have no objections.

LGTM

File: src/Microsoft.Framework.PackageManager/Scripts/ScriptExecutor.cs [hunk 1]
```diff
@@ -97,9 +97,9 @@ public bool Execute(Runtime.Project project, string scriptName, Func<string, str
                         scriptArguments[0] = $"{ quote }{ prefix }{ scriptCandidate }{ quote }";
                     }
 
-                    // Always use /bin/bash -c in order to support redirection and so on; similar to Windows case.
+                    // Always use /usr/bin/env bash -c in order to support redirection and so on; similar to Windows case.
                     // Unlike Windows, must escape quotation marks within the newly-quoted string.
-                    scriptArguments = new[] { "/bin/bash", "-c", "\"" }
+                    scriptArguments = new[] { "/usr/bin/env bash", "-c", "\"" }
```
Comment (dougbu) [inline src/Microsoft.Framework.PackageManager/Scripts/ScriptExecutor.cs]: First element in array needs to be split in two. `/usr/bin/env` is the executable while `bash` is the first argument.

Comment (jasonwilliams200OK) [inline src/Microsoft.Framework.PackageManager/Scripts/ScriptExecutor.cs]: Fixed by e0cdbdd. Thanks! :)

**Comment by dougbu** [comment]
@josteink the Travis build failed because `dnu restore` failed, which it sometimes does on Linux.

@jasonwilliams200OK am I correct #1756 can be closed? Looks like this PR changes every file that one does (did).

**Comment by jasonwilliams200OK** [comment]
@dougbu, yes this supersedes #1756.

**Comment by dougbu** [comment]
@jasonmalinowski this LGTM but not to Travis. I suspect something (that I'm missing) is off with the changes in `ScriptExecutor` or `DnuRestoreTests`. Probably best to build on some Linux system and see exactly what works and what doesn't.

```
Restoring packages for /tmp/5bp56xqm.dd2/Project Name/project.json
Executing script 'prerestore' in project.json
/usr/bin/env: bash --restricted: No such file or directory
The 'prerestore' script failed with status code 127.
Errors in prerestore
    The 'prerestore' script failed with status code 127.
   Microsoft.Framework.PackageManager.DnuRestoreTests.DnuRestore_ExecutesScripts(flavor: "mono", os: null, architecture: null) [FAIL]
      Assert.Equal() Failure
      Expected: 0
      Actual:   1
      Stack Trace:
           at Microsoft.Framework.PackageManager.DnuRestoreTests.DnuRestore_ExecutesScripts (System.String flavor, System.String os, System.String architecture) [0x00000] in <filename unknown>:0 
...
```

**Comment by dougbu** [comment]
FYI this PR covers some build issues and some product issues. But even after this and changing `build.sh` in all aspnet repos, we will not have a product working end-to-end on FreeBSD.

The `node.js` situation looks inconsistent at best. For example a number of the modules `gulp` and `grunt` depend on (say `dateformat`) also use `/bin/bash`.

**Comment by josteink** [comment]
I don't have an issue with that. Right now dnu is the one stopping corefx from building. Getting full support is probably a bigger endeavour and needs to be taken one step at a time.

**Comment by davidfowl** [comment]
Still it breaks travis which means it'll likely break real systems. Why is hardcoding `/usr/bin/env` better?

**Comment by jasonmalinowski** [comment]
@dougbu, did you mean to mention me? I was recently added to the ASP.NET org and thus I might have hijacked your previous Jason-auto-complete.

**Comment by davidfowl** [comment]
@jasonmalinowski nope he meant @jasonwilliams200OK

**Comment by jasonwilliams200OK** [comment]
> Still it breaks travis which means it'll likely break real systems. Why is hardcoding  /usr/bin/env  better?

@davidfowl, yup it is only one test failing on line 8500 something. However, it passes all tests with this change:

```diff
diff --git a/test/Microsoft.Framework.PackageManager.FunctionalTests/DnuRestoreTests.cs b/test/Microsoft.Framework.PackageManager.FunctionalTests/DnuRestoreTests.cs
index d595847..646677a 100644
--- a/test/Microsoft.Framework.PackageManager.FunctionalTests/DnuRestoreTests.cs
+++ b/test/Microsoft.Framework.PackageManager.FunctionalTests/DnuRestoreTests.cs
@@ -101,7 +101,7 @@ goto argumentStart
   }
 }";
                 scriptContent =
-@"#!/usr/bin/env bash --restricted
+@"#!/usr/bin/env bash
 set -o errexit
 
 for arg in ""$@""; do
```

On the flipside, it is preventing use building CoreFX against FreeBSD. Note that we have already made these changes in `.sh` scripts located in Roslyn, CoreFX and CoreCLR repositories. The remaining pieces are aspnet:dnx, aspnet:dnvm, aspnet:Universe and aspnet:Home. I have opened the ditto PRs in all of them. o_0

Should I make that change and leave a `TODO` there? Later someone with bash expertise can figure out how to deal with this `--restricted`-mode glitch. :)

**Comment by dougbu** [comment]
@jasonwilliams200OK (yup, that's the Jason I meant :smiley_cat:) Do not disable this test. It's one of the few that exercises script execution. Without it we may hit failures on the _second_ build after your change goes in.

I suggest removing `--restricted` from that embedded script. I included that for some definition of "safety", not because it was mandatory.

**Comment by dougbu** [comment]
Oh, I'm agreeing with you. Good.

**Comment by jasonwilliams200OK** [comment]
Thanks @dougbu! Its done. :metal:

**Comment by jasonwilliams200OK** [comment]
That was some interrupt. It built on my fork: https://travis-ci.org/jasonwilliams200OK/dnx/builds/66921589. I have re-pushed, lets see.. :)

**Comment by dougbu** [comment]
@jasonwilliams200OK Travis is failing for reasons completely unrelated to this PR. Wait a bit while our `dev` branches settle down. @davidfowl or I will let you know...

**Comment by jasonwilliams200OK** [comment]
@dougbu, @davidfowl, seems like you guys have fixed the issue in some dependency. :smile: 
TravisCI build succeeded! :tada:

**Comment by dougbu** [comment]
Yeah, that happened finally earlier today. Sorry we didn't let you know.

I'll merge i.e. :shipit: this for you.

**Comment by dougbu** [comment]
2c3d535

**Comment by dougbu** [comment]
Many thanks @jasonwilliams200OK

**Comment by dougbu** [comment]
@jasonwilliams200OK @davidfowl do we have an all-up FreeBSD or bash location issue in one of the repos? Hoping we're tracking making the `build.sh` changes in all the Universe repos somewhere...

**Comment by jasonwilliams200OK** [comment]
@dougbu, I opened PRs in all repositories :)

aspnet/dnx#2066
aspnet/dnvm#295
aspnet/Universe#238
aspnet/Home#671

**Comment by dougbu** [comment]
@jasonwilliams200OK great, those PRs cover the basics. The mundane part is about updating `build.sh` in aspnet/AntiForgery, aspnet/WebSockets, and everything in-between. And an issue might help track the four main PRs 'til they are all closed.

Note we have scripts to update most of the repos. Still mundane but less so. Just need to apply those scripts over the next few days...

**Pull request closed without merging**